scripts/affiliate_server.py

```diff
@@ -171,30 +171,52 @@ def get_pending_books(books):
     return pending_books
 
 
-def process_amazon_batch(isbn_10s: list[str]) -> None:
+def make_cache_key(product: dict[str, Any]) -> str:
+    """
+    Takes a `product` returned from `vendor.get_products()` and returns a cache key to
+    identify the product. For a given product, the cache key will be either (1) its
+    ISBN 13, or (2) it's non-ISBN 10 ASIN (i.e. one that starts with `B`).
+    """
+    if (isbn_13s := product.get("isbn_13")) and len(isbn_13s):
+        return isbn_13s[0]
+
+    if product.get("isbn_10") and (
+        cache_key := isbn_10_to_isbn_13(product.get("isbn_10", [])[0])
+    ):
+        return cache_key
+
+    if (source_records := product.get("source_records")) and (
+        amazon_record := next(
+            (record for record in source_records if record.startswith("amazon:")), ""
+        )
+    ):
+        return amazon_record.split(":")[1]
+
+    return ""
+
+
+def process_amazon_batch(isbn_10s_or_asins: list[str]) -> None:
     """
     Call the Amazon API to get the products for a list of isbn_10s and store
     each product in memcache using amazon_product_{isbn_13} as the cache key.
     """
-    logger.info(f"process_amazon_batch(): {len(isbn_10s)} items")
+    logger.info(f"process_amazon_batch(): {len(isbn_10s_or_asins)} items")
```

Comment:
Some of the Amazon specific terminology in the comments and naming may be a bit confusing, an ASIN is a 10-digit identifier which is often an ISBN-10.  Variable names like `isbn_10s_or_asins` could simply be `asin` if that's what the Amazon API is returning, but from the surrounding context I'm not sure.

Referring to all ASINs as ISBN-10 is going to be wrong sometimes (unless they are pre-filtered ASINs), but calling them ISBN-10s or ASINs is either redundant, or implies that some of the values are ISBN-10s which do not correspond to ASINs on Amazon.com, but I don't think that's what's happening in this module?

I thought the Amazon API was clear with its terminology and distinguishing between ISBNs and ASINs?

Reply:
I will try to fix up the nomenclature. I agree it's not going to win awards. :)